### ISSUE TITLE
Add support for unnesting collections of expressions with DSL::unnest

### ISSUE THREAD

**Issue by atollk** [jOOQ/jOOQ#13375]
### Expected behavior

```kotlin
DSL.using(DefaultConfiguration(SQLDialect.POSTGRES)).selectFrom(DSL.unnest(listOf(DSL.value(1)))).hashCode()
```

should result in some hash value.

### Actual behavior

```
Type class org.jooq.impl.Val is not supported in dialect null
org.jooq.exception.SQLDialectNotSupportedException: Type class org.jooq.impl.Val is not supported in dialect null
	at org.jooq_3.16.5.POSTGRES.debug(Unknown Source)
	at app//org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:811)
	at app//org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:755)
	at app//org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:765)
	at app//org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:755)
	at app//org.jooq.impl.AbstractField.cast(AbstractField.java:259)
	at app//org.jooq.impl.DSL.cast(DSL.java:15438)
	at app//org.jooq.impl.DefaultBinding$DefaultArrayBinding.sqlInline0(DefaultBinding.java:1184)
	at app//org.jooq.impl.DefaultBinding$DefaultArrayBinding.sqlInline0(DefaultBinding.java:1158)
	at app//org.jooq.impl.DefaultBinding$InternalBinding.sql(DefaultBinding.java:952)
	at app//org.jooq.impl.DefaultBinding$InternalBinding.sql(DefaultBinding.java:944)
	at app//org.jooq.impl.Val.accept(Val.java:186)
	at app//org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:721)
	at app//org.jooq.impl.AbstractContext.visit(AbstractContext.java:295)
	at app//org.jooq.impl.ArrayTable$PostgresHSQLDBTable.accept(ArrayTable.java:208)
	at app//org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:721)
	at app//org.jooq.impl.AbstractContext.visit(AbstractContext.java:255)
	at app//org.jooq.impl.Alias.toSQLWrapped(Alias.java:381)
	at app//org.jooq.impl.Alias.toSQLWrapped(Alias.java:376)
	at app//org.jooq.impl.Alias.acceptDeclareAliasStandard(Alias.java:297)
	at app//org.jooq.impl.Alias.accept(Alias.java:183)
	at app//org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:721)
	at app//org.jooq.impl.AbstractContext.visit(AbstractContext.java:295)
	at app//org.jooq.impl.TableAlias.accept(TableAlias.java:137)
	at app//org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:721)
	at app//org.jooq.impl.AbstractContext.visit(AbstractContext.java:295)
	at app//org.jooq.impl.ArrayTable.accept(ArrayTable.java:161)
	at app//org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:721)
	at app//org.jooq.impl.AbstractContext.visit(AbstractContext.java:295)
	at app//org.jooq.impl.QueryPartCollectionView.acceptElement(QueryPartCollectionView.java:243)
	at app//org.jooq.impl.TableList.acceptElement(TableList.java:92)
	at app//org.jooq.impl.TableList.acceptElement(TableList.java:63)
	at app//org.jooq.impl.QueryPartCollectionView.accept(QueryPartCollectionView.java:221)
	at app//org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:721)
	at app//org.jooq.impl.AbstractContext.visit(AbstractContext.java:295)
	at app//org.jooq.impl.UpdateQueryImpl.lambda$acceptFrom$1(UpdateQueryImpl.java:776)
	at app//org.jooq.impl.AbstractContext.toggle(AbstractContext.java:324)
	at app//org.jooq.impl.AbstractContext.declareTables(AbstractContext.java:563)
	at app//org.jooq.impl.UpdateQueryImpl.acceptFrom(UpdateQueryImpl.java:776)
	at app//org.jooq.impl.UpdateQueryImpl.accept0(UpdateQueryImpl.java:685)
	at app//org.jooq.impl.AbstractDMLQuery.accept(AbstractDMLQuery.java:642)
	at app//org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:721)
	at app//org.jooq.impl.AbstractContext.visit(AbstractContext.java:295)
	at app//org.jooq.impl.DefaultDSLContext.renderInlined(DefaultDSLContext.java:725)
	at app//org.jooq.impl.AbstractQueryPart.hashCode(AbstractQueryPart.java:183)
``` 

### Steps to reproduce the problem

```kotlin
DSL.using(DefaultConfiguration(SQLDialect.POSTGRES)).selectFrom(DSL.unnest(listOf(DSL.value(1)))).hashCode()
```

Removing the configuration part fixes the issue, i.e.

```kotlin
DSL.selectFrom(DSL.unnest(listOf(DSL.value(1)))).hashCode()
```

produces a value without error.

### Versions

- jOOQ: 3.16.5
- Kotlin: 1.5
- Database (include vendor): Postgres
- OS: Arch Linux

**Comment by lukaseder** [comment]
Thanks a lot for your report. Indeed, it's currently not possible to pass a list (or array) of `Field` expressions, only of actual bind values, e.g. `listOf(1)` should work.

**Comment by lukaseder** [comment]
Looking into this now. I can reproduce the `hashCode()` problem, but it seems that the query also doesn't work when passing `listOf(1)`, the error being:

```
org.jooq.exception.DataAccessException: SQL [select ?, "COLUMN_VALUE" from unnest(cast(? as any[])) as "x" ("COLUMN_VALUE") -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT]; ERROR: syntax error at or near "any"
  Position: 50
	at org.jooq_3.18.0-SNAPSHOT.POSTGRES.debug(Unknown Source) ~[?:?]
	at org.jooq.impl.Tools.translate(Tools.java:3363) ~[classes/:?]
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:686) ~[classes/:?]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:355) ~[classes/:?]
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:290) ~[classes/:?]
	at org.jooq.impl.SelectImpl.fetch(SelectImpl.java:2837) ~[classes/:?]
	at org.jooq.test.all.testcases.TableFunctionTests.testUnnest0(TableFunctionTests.java:286) ~[test-classes/:?]
	at org.jooq.test.all.testcases.TableFunctionTests.testUnnest(TableFunctionTests.java:255) ~[test-classes/:?]
	at org.jooq.test.jOOQAbstractTest.testUnnest(jOOQAbstractTest.java:8753) ~[test-classes/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413) ~[junit-4.13.2.jar:4.13.2]
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93) ~[.cp/:?]
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40) ~[.cp/:?]
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529) ~[.cp/:?]
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756) ~[.cp/:?]
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452) ~[.cp/:?]
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210) ~[.cp/:?]
Caused by: org.postgresql.util.PSQLException: ERROR: syntax error at or near "any"
  Position: 50
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2676) ~[postgresql-42.4.1.jar:42.4.1]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2366) ~[postgresql-42.4.1.jar:42.4.1]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:356) ~[postgresql-42.4.1.jar:42.4.1]
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490) ~[postgresql-42.4.1.jar:42.4.1]
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408) ~[postgresql-42.4.1.jar:42.4.1]
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:181) ~[postgresql-42.4.1.jar:42.4.1]
	at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:170) ~[postgresql-42.4.1.jar:42.4.1]
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:219) ~[classes/:?]
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:219) ~[classes/:?]
	at org.jooq.impl.Tools.executeStatementAndGetFirstResultSet(Tools.java:4619) ~[classes/:?]
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:236) ~[classes/:?]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:341) ~[classes/:?]
	... 37 more
```

The problem is that the array type is unknown due to generic type erasure, so jOOQ generates a `cast(? as any[])`, which doesn't work. Possibly, that's related

**Comment by lukaseder** [comment]
The root cause here is one of API design. We're just accepting `Collection<?>` or `Object[]`, assuming these are going to be the values contained in a single array bind value.

In your case, however, the array is a different one. Instead of `CAST(? AS INT[])`, the result should be `ARRAY[?, ?, ..., ?]::INT[]`, for example.

**Comment by lukaseder** [comment]
> Looking into this now. I can reproduce the `hashCode()` problem, but it seems that the query also doesn't work when passing `listOf(1)`, the error being: [...]

The problems are related, but not the same. I've created a separate feature request: https://github.com/jOOQ/jOOQ/issues/14074

This issue here will only fix the problem you've reported.

**Comment by lukaseder** [comment]
The new feature can be emulated on dialects without array support using the existing `ArrayTableEmulation` with just a few tweaks, producing a derived table with `UNION ALL` queries.

**Comment by lukaseder** [comment]
Implemented in jOOQ 3.18.0
